# **New York City Yellow Taxi Data**

## Objective
In this case study you will be learning exploratory data analysis (EDA) with the help of a dataset on yellow taxi rides in New York City. This will enable you to understand why EDA is an important step in the process of data science and machine learning.

## **Problem Statement**
As an analyst at an upcoming taxi operation in NYC, you are tasked to use the 2023 taxi trip data to uncover insights that could help optimise taxi operations. The goal is to analyse patterns in the data that can inform strategic decisions to improve service efficiency, maximise revenue, and enhance passenger experience.

## Tasks
You need to perform the following steps for successfully completing this assignment:
1. Data Loading
2. Data Cleaning
3. Exploratory Analysis: Bivariate and Multivariate
4. Creating Visualisations to Support the Analysis
5. Deriving Insights and Stating Conclusions

---

**NOTE:** The marks given along with headings and sub-headings are cumulative marks for those particular headings/sub-headings.<br>

The actual marks for each task are specified within the tasks themselves.

For example, marks given with heading *2* or sub-heading *2.1* are the cumulative marks, for your reference only. <br>

The marks you will receive for completing tasks are given with the tasks.

Suppose the marks for two tasks are: 3 marks for 2.1.1 and 2 marks for 3.2.2, or
* 2.1.1 [3 marks]
* 3.2.2 [2 marks]

then, you will earn 3 marks for completing task 2.1.1 and 2 marks for completing task 3.2.2.


---

## Data Understanding
The yellow taxi trip records include fields capturing pick-up and drop-off dates/times, pick-up and drop-off locations, trip distances, itemized fares, rate types, payment types, and driver-reported passenger counts.

The data is stored in Parquet format (*.parquet*). The dataset is from 2009 to 2024. However, for this assignment, we will only be using the data from 2023.

The data for each month is present in a different parquet file. You will get twelve files for each of the months in 2023.

The data was collected and provided to the NYC Taxi and Limousine Commission (TLC) by technology providers like vendors and taxi hailing apps. <br>

You can find the link to the TLC trip records page here: https://www.nyc.gov/site/tlc/about/tlc-trip-record-data.page

###  Data Description
You can find the data description here: [Data Dictionary](https://www.nyc.gov/assets/tlc/downloads/pdf/data_dictionary_trip_records_yellow.pdf)

**Trip Records**



|Field Name       |description |
|:----------------|:-----------|
| VendorID | A code indicating the TPEP provider that provided the record. <br> 1= Creative Mobile Technologies, LLC; <br> 2= VeriFone Inc. |
| tpep_pickup_datetime | The date and time when the meter was engaged.  |
| tpep_dropoff_datetime | The date and time when the meter was disengaged.   |
| Passenger_count | The number of passengers in the vehicle. <br> This is a driver-entered value. |
| Trip_distance | The elapsed trip distance in miles reported by the taximeter. |
| PULocationID | TLC Taxi Zone in which the taximeter was engaged |
| DOLocationID | TLC Taxi Zone in which the taximeter was disengaged |
|RateCodeID |The final rate code in effect at the end of the trip.<br> 1 = Standard rate <br> 2 = JFK <br> 3 = Newark <br>4 = Nassau or Westchester <br>5 = Negotiated fare <br>6 = Group ride |
|Store_and_fwd_flag |This flag indicates whether the trip record was held in vehicle memory before sending to the vendor, aka “store and forward,” because the vehicle did not have a connection to the server.  <br>Y= store and forward trip <br>N= not a store and forward trip |
|Payment_type| A numeric code signifying how the passenger paid for the trip. <br> 1 = Credit card <br>2 = Cash <br>3 = No charge <br>4 = Dispute <br>5 = Unknown <br>6 = Voided trip |
|Fare_amount| The time-and-distance fare calculated by the meter. <br>Extra Miscellaneous extras and surcharges.  Currently, this only includes the 0.50 and 1 USD rush hour and overnight charges. |
|MTA_tax |0.50 USD MTA tax that is automatically triggered based on the metered rate in use. |
|Improvement_surcharge | 0.30 USD improvement surcharge assessed trips at the flag drop. The improvement surcharge began being levied in 2015. |
|Tip_amount |Tip amount – This field is automatically populated for credit card tips. Cash tips are not included. |
| Tolls_amount | Total amount of all tolls paid in trip.  |
| total_amount | The total amount charged to passengers. Does not include cash tips. |
|Congestion_Surcharge |Total amount collected in trip for NYS congestion surcharge. |
| Airport_fee | 1.25 USD for pick up only at LaGuardia and John F. Kennedy Airports|

Although the amounts of extra charges and taxes applied are specified in the data dictionary, you will see that some cases have different values of these charges in the actual data.

**Taxi Zones**

Each of the trip records contains a field corresponding to the location of the pickup or drop-off of the trip, populated by numbers ranging from 1-263.

These numbers correspond to taxi zones, which may be downloaded as a table or map/shapefile and matched to the trip records using a join.

This is covered in more detail in later sections.

---

## **1** Data Preparation

<font color = red>[5 marks]</font> <br>

### Import Libraries

In [2]:
# Import warnings

import warnings

warnings.filterwarnings("ignore")

In [3]:
# Import the libraries you will be using for analysis

In [1]:
# Recommended versions
# numpy version: 1.26.4
# pandas version: 2.2.2
# matplotlib version: 3.10.0
# seaborn version: 0.13.2
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

# Check versions
print("numpy version:", np.__version__)
print("pandas version:", pd.__version__)
print("matplotlib version:", plt.matplotlib.__version__)
print("seaborn version:", sns.__version__)

numpy version: 1.26.4
pandas version: 2.2.2
matplotlib version: 3.10.0
seaborn version: 0.13.2


### **1.1** Load the dataset
<font color = red>[5 marks]</font> <br>

You will see twelve files, one for each month.

To read parquet files with Pandas, you have to follow a similar syntax as that for CSV files.

`df = pd.read_parquet('file.parquet')`

In [5]:
# Try loading one file
df = pd.read_parquet("trip_records/2023-1.parquet")
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3041714 entries, 0 to 3066765
Data columns (total 19 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int64         
 1   tpep_pickup_datetime   datetime64[us]
 2   tpep_dropoff_datetime  datetime64[us]
 3   passenger_count        float64       
 4   trip_distance          float64       
 5   RatecodeID             float64       
 6   store_and_fwd_flag     object        
 7   PULocationID           int64         
 8   DOLocationID           int64         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge  float64       
 16  total_amount           float64       
 17  congestion_surcharge   float64       
 18  airport_fee            floa

In [6]:
df.shape

(3041714, 19)

In [7]:
df

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,airport_fee
0,2,2023-01-01 00:32:10,2023-01-01 00:40:36,1.0,0.97,1.0,N,161,141,2,9.30,1.00,0.5,0.00,0.0,1.0,14.30,2.5,0.00
1,2,2023-01-01 00:55:08,2023-01-01 01:01:27,1.0,1.10,1.0,N,43,237,1,7.90,1.00,0.5,4.00,0.0,1.0,16.90,2.5,0.00
2,2,2023-01-01 00:25:04,2023-01-01 00:37:49,1.0,2.51,1.0,N,48,238,1,14.90,1.00,0.5,15.00,0.0,1.0,34.90,2.5,0.00
3,1,2023-01-01 00:03:48,2023-01-01 00:13:25,0.0,1.90,1.0,N,138,7,1,12.10,7.25,0.5,0.00,0.0,1.0,20.85,0.0,1.25
4,2,2023-01-01 00:10:29,2023-01-01 00:21:19,1.0,1.43,1.0,N,107,79,1,11.40,1.00,0.5,3.28,0.0,1.0,19.68,2.5,0.00
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3066761,2,2023-01-31 23:58:34,2023-02-01 00:12:33,NaN,3.05,NaN,None,107,48,0,15.80,0.00,0.5,3.96,0.0,1.0,23.76,NaN,NaN
3066762,2,2023-01-31 23:31:09,2023-01-31 23:50:36,NaN,5.80,NaN,None,112,75,0,22.43,0.00,0.5,2.64,0.0,1.0,29.07,NaN,NaN
3066763,2,2023-01-31 23:01:05,2023-01-31 23:25:36,NaN,4.67,NaN,None,114,239,0,17.61,0.00,0.5,5.32,0.0,1.0,26.93,NaN,NaN
3066764,2,2023-01-31 23:40:00,2023-01-31 23:53:00,NaN,3.15,NaN,None,230,79,0,18.15,0.00,0.5,4.43,0.0,1.0,26.58,NaN,NaN


In [8]:
df.isnull().sum()

VendorID                     0
tpep_pickup_datetime         0
tpep_dropoff_datetime        0
passenger_count          71686
trip_distance                0
RatecodeID               71686
store_and_fwd_flag       71686
PULocationID                 0
DOLocationID                 0
payment_type                 0
fare_amount                  0
extra                        0
mta_tax                      0
tip_amount                   0
tolls_amount                 0
improvement_surcharge        0
total_amount                 0
congestion_surcharge     71686
airport_fee              71686
dtype: int64

In [9]:
df.to_csv("2023-1.csv")

In [10]:
df["tpep_pickup_datetime"].dt.hour.value_counts()

tpep_pickup_datetime
18    214294
17    207902
15    194886
16    194267
19    191289
14    190067
13    177379
12    168642
20    164436
21    160194
11    153084
22    146069
10    142658
9     130288
8     116209
23    113398
7      86313
0      84044
1      59165
6      43410
2      41508
3      27011
5      17724
4      17477
Name: count, dtype: int64

In [11]:
df.dtypes

VendorID                          int64
tpep_pickup_datetime     datetime64[us]
tpep_dropoff_datetime    datetime64[us]
passenger_count                 float64
trip_distance                   float64
RatecodeID                      float64
store_and_fwd_flag               object
PULocationID                      int64
DOLocationID                      int64
payment_type                      int64
fare_amount                     float64
extra                           float64
mta_tax                         float64
tip_amount                      float64
tolls_amount                    float64
improvement_surcharge           float64
total_amount                    float64
congestion_surcharge            float64
airport_fee                     float64
dtype: object

In [12]:
# df['tpep_dropoff_datetime'].dt.month.value_counts()
df["hour"] = df["tpep_pickup_datetime"].dt.hour

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3041714 entries, 0 to 3066765
Data columns (total 20 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int64         
 1   tpep_pickup_datetime   datetime64[us]
 2   tpep_dropoff_datetime  datetime64[us]
 3   passenger_count        float64       
 4   trip_distance          float64       
 5   RatecodeID             float64       
 6   store_and_fwd_flag     object        
 7   PULocationID           int64         
 8   DOLocationID           int64         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge  float64       
 16  total_amount           float64       
 17  congestion_surcharge   float64       
 18  airport_fee            floa

In [14]:
df["hour"].value_counts()

hour
18    214294
17    207902
15    194886
16    194267
19    191289
14    190067
13    177379
12    168642
20    164436
21    160194
11    153084
22    146069
10    142658
9     130288
8     116209
23    113398
7      86313
0      84044
1      59165
6      43410
2      41508
3      27011
5      17724
4      17477
Name: count, dtype: int64

In [15]:
hour_data = df.copy()
hour_data["hour"].value_counts()

hour
18    214294
17    207902
15    194886
16    194267
19    191289
14    190067
13    177379
12    168642
20    164436
21    160194
11    153084
22    146069
10    142658
9     130288
8     116209
23    113398
7      86313
0      84044
1      59165
6      43410
2      41508
3      27011
5      17724
4      17477
Name: count, dtype: int64

In [16]:
hour_data[hour_data["hour"] == 0]

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,airport_fee,hour
0,2,2023-01-01 00:32:10,2023-01-01 00:40:36,1.0,0.97,1.0,N,161,141,2,9.30,1.00,0.5,0.00,0.0,1.0,14.30,2.5,0.00,0
1,2,2023-01-01 00:55:08,2023-01-01 01:01:27,1.0,1.10,1.0,N,43,237,1,7.90,1.00,0.5,4.00,0.0,1.0,16.90,2.5,0.00,0
2,2,2023-01-01 00:25:04,2023-01-01 00:37:49,1.0,2.51,1.0,N,48,238,1,14.90,1.00,0.5,15.00,0.0,1.0,34.90,2.5,0.00,0
3,1,2023-01-01 00:03:48,2023-01-01 00:13:25,0.0,1.90,1.0,N,138,7,1,12.10,7.25,0.5,0.00,0.0,1.0,20.85,0.0,1.25,0
4,2,2023-01-01 00:10:29,2023-01-01 00:21:19,1.0,1.43,1.0,N,107,79,1,11.40,1.00,0.5,3.28,0.0,1.0,19.68,2.5,0.00,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3064418,2,2023-01-31 00:20:05,2023-01-31 00:40:33,NaN,5.16,NaN,None,158,43,0,20.05,0.00,0.5,2.40,0.0,1.0,26.45,NaN,NaN,0
3064419,2,2023-01-31 00:40:17,2023-01-31 00:44:43,NaN,1.04,NaN,None,48,143,0,14.40,0.00,0.5,0.92,0.0,1.0,19.32,NaN,NaN,0
3064420,1,2023-01-31 00:17:44,2023-01-31 00:41:38,NaN,0.00,NaN,None,107,166,0,13.00,1.00,0.5,5.00,0.0,1.0,25.00,NaN,NaN,0
3064421,2,2023-01-31 00:30:00,2023-01-31 00:38:00,NaN,1.70,NaN,None,137,90,0,12.47,0.00,0.5,2.33,0.0,1.0,18.80,NaN,NaN,0


In [20]:
hour_data["tpep_pickup_datetime"].dt.day_of_year.value_counts()

tpep_pickup_datetime
26     114013
19     113775
14     111960
21     111051
27     110757
28     110681
12     110571
13     109802
20     108566
25     108366
18     105333
11     105333
7      104114
24     103051
6      101727
5      100269
17     100242
31      99654
10      99158
15      96046
4       94290
23      89039
22      88903
29      87403
3       85011
9       84628
8       84337
30      83069
16      79406
1       76021
2       65090
365        25
32         10
298         7
297         4
366         2
Name: count, dtype: int64

In [22]:
hour_data["tpep_pickup_datetime"].dt.year.value_counts()

tpep_pickup_datetime
2023    3041676
2022         36
2008          2
Name: count, dtype: int64

In [18]:
hour_data[
    (hour_data["tpep_pickup_datetime"].dt.day_of_year == 1)
    & (hour_data["tpep_pickup_datetime"].dt.hour == 0)
]

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,airport_fee,hour
0,2,2023-01-01 00:32:10,2023-01-01 00:40:36,1.0,0.97,1.0,N,161,141,2,9.30,1.00,0.5,0.00,0.0,1.0,14.30,2.5,0.00,0
1,2,2023-01-01 00:55:08,2023-01-01 01:01:27,1.0,1.10,1.0,N,43,237,1,7.90,1.00,0.5,4.00,0.0,1.0,16.90,2.5,0.00,0
2,2,2023-01-01 00:25:04,2023-01-01 00:37:49,1.0,2.51,1.0,N,48,238,1,14.90,1.00,0.5,15.00,0.0,1.0,34.90,2.5,0.00,0
3,1,2023-01-01 00:03:48,2023-01-01 00:13:25,0.0,1.90,1.0,N,138,7,1,12.10,7.25,0.5,0.00,0.0,1.0,20.85,0.0,1.25,0
4,2,2023-01-01 00:10:29,2023-01-01 00:21:19,1.0,1.43,1.0,N,107,79,1,11.40,1.00,0.5,3.28,0.0,1.0,19.68,2.5,0.00,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2995373,2,2023-01-01 00:41:41,2023-01-01 01:06:07,NaN,8.74,NaN,None,220,24,0,35.44,0.00,0.5,7.39,0.0,1.0,44.33,NaN,NaN,0
2995374,2,2023-01-01 00:47:00,2023-01-01 00:55:00,NaN,1.92,NaN,None,236,142,0,13.29,0.00,0.5,2.59,0.0,1.0,19.88,NaN,NaN,0
2995375,2,2023-01-01 00:41:00,2023-01-01 01:23:00,NaN,6.82,NaN,None,231,151,0,35.28,0.00,0.5,7.86,0.0,1.0,47.14,NaN,NaN,0
2995376,1,2023-01-01 00:31:02,2023-01-01 00:41:49,NaN,0.00,NaN,None,238,263,0,10.00,1.00,0.5,2.25,0.0,1.0,19.25,NaN,NaN,0


In [19]:
hour_data[
    (hour_data["tpep_pickup_datetime"].dt.day_of_year == 1)
    & (hour_data["tpep_pickup_datetime"].dt.hour == 2)
]

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,airport_fee,hour
6080,2,2023-01-01 02:00:30,2023-01-01 02:07:20,3.0,1.50,1.0,N,238,166,1,10.00,1.0,0.5,0.75,0.0,1.0,15.75,2.5,0.0,2
6171,2,2023-01-01 02:02:25,2023-01-01 02:06:57,1.0,0.44,1.0,N,231,209,1,6.50,1.0,0.5,1.88,0.0,1.0,13.38,2.5,0.0,2
7171,2,2023-01-01 02:00:08,2023-01-01 02:09:51,2.0,1.99,1.0,N,161,236,1,12.10,1.0,0.5,3.42,0.0,1.0,20.52,2.5,0.0,2
7274,2,2023-01-01 02:03:32,2023-01-01 02:07:55,6.0,0.63,1.0,N,234,249,1,6.50,1.0,0.5,2.30,0.0,1.0,13.80,2.5,0.0,2
7516,2,2023-01-01 02:06:20,2023-01-01 02:12:24,1.0,1.31,1.0,N,141,263,1,8.60,1.0,0.5,3.40,0.0,1.0,17.00,2.5,0.0,2
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2996371,2,2023-01-01 02:06:14,2023-01-01 02:17:59,NaN,3.13,NaN,None,236,166,0,19.63,0.0,0.5,0.00,0.0,1.0,23.63,NaN,NaN,2
2996372,2,2023-01-01 02:21:37,2023-01-01 02:41:02,NaN,4.35,NaN,None,40,107,0,26.13,0.0,0.5,6.03,0.0,1.0,36.16,NaN,NaN,2
2996373,2,2023-01-01 02:25:32,2023-01-01 02:37:44,NaN,3.28,NaN,None,148,162,0,19.83,0.0,0.5,0.00,0.0,1.0,23.83,NaN,NaN,2
2996374,2,2023-01-01 02:42:37,2023-01-01 03:10:28,NaN,7.08,NaN,None,255,141,0,33.68,0.0,0.5,7.54,0.0,1.0,45.22,NaN,NaN,2


In [ ]:
# todo
def concat_file(hour_data):
    """concats hourly data of each file
        taking only 5% data of each hour
    Args:
        file (parquet): Monthly data of Problem Set
    """
    print(hour_data)
    hour_data = pd.read_parquet(hour_data)
    hour_data["hour"] = hour_data["tpep_pickup_datetime"].dt.hour
    hour_data["date"] = hour_data["tpep_pickup_datetime"].dt.day_of_year
    max_date = max(hour_data["date"])
    sample = hour_data[
        (hour_data["tpep_pickup_datetime"].dt.day_of_year == 1)
        & (hour_data["tpep_pickup_datetime"].dt.hour == 0)
    ]
    total = pd.DataFrame()
    for hour in range(1, 24):
        sample = pd.concat(
            [
                sample,
                hour_data[
                    (hour_data["tpep_pickup_datetime"].dt.day_of_year == hour)
                    & (hour_data["tpep_pickup_datetime"].dt.hour == hour)
                ],
            ]
        )

    for date in range(2, max_date):
        sample = pd.concat([])

        pass
    pass

    # for hour in total['hour']:
    #     sample = total[total['hour']==hour].sample(frac=0.05,random_state=42)
    #     sampled_data = pd.concat([sampled_data,sample])
    # return total

In [108]:
check = concat_file("trip_records/2023-1.parquet")

trip_records/2023-1.parquet


KeyboardInterrupt: 

In [66]:
total.shape

(152083, 20)

In [67]:
total

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,airport_fee,hour
784218,2,2023-01-10 00:44:09,2023-01-10 00:51:23,1.0,1.97,1.0,N,142,164,1,10.7,1.00,0.5,3.14,0.0,1.0,18.84,2.5,0.00,0
1316921,1,2023-01-15 00:32:10,2023-01-15 00:42:27,1.0,1.50,1.0,N,113,79,1,10.0,3.50,0.5,2.00,0.0,1.0,17.00,2.5,0.00,0
2732390,1,2023-01-29 00:29:35,2023-01-29 00:42:50,2.0,2.60,1.0,N,263,143,1,15.6,3.50,0.5,4.10,0.0,1.0,24.70,2.5,0.00,0
1203249,2,2023-01-14 00:14:30,2023-01-14 00:27:04,1.0,1.79,1.0,N,79,158,1,13.5,1.00,0.5,1.48,0.0,1.0,19.98,2.5,0.00,0
137926,2,2023-01-03 00:55:05,2023-01-03 01:03:28,1.0,3.34,1.0,N,132,132,2,15.6,1.00,0.5,0.00,0.0,1.0,19.35,0.0,1.25,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1904739,2,2023-01-20 23:21:31,2023-01-20 23:28:40,1.0,0.98,1.0,N,249,113,2,8.6,1.00,0.5,0.00,0.0,1.0,13.60,2.5,0.00,23
71943,2,2023-01-01 23:16:42,2023-01-01 23:45:50,1.0,11.66,1.0,N,264,37,1,48.5,1.00,0.5,3.74,0.0,1.0,57.24,2.5,0.00,23
2397817,1,2023-01-25 23:52:02,2023-01-26 00:02:31,4.0,5.10,1.0,N,132,216,2,21.9,2.25,0.5,0.00,0.0,1.0,25.65,0.0,1.25,23
1483332,2,2023-01-16 23:50:01,2023-01-16 23:54:26,1.0,1.83,1.0,N,74,263,1,9.3,1.00,0.5,2.86,0.0,1.0,17.16,2.5,0.00,23


In [68]:
total.shape[0] / df.shape[0]

0.04999911234258053

In [88]:
import os

os.listdir("trip_records")

['2023-12.parquet',
 '2023-6.parquet',
 '2023-7.parquet',
 '2023-5.parquet',
 '2023-11.parquet',
 '2023-10.parquet',
 '2023-4.parquet',
 '2023-1.parquet',
 '2023-8.parquet',
 '2023-9.parquet',
 '2023-2.parquet',
 '2023-3.parquet']

In [89]:
sampled_data = pd.DataFrame()
for file in os.listdir("trip_records"):
    sample = concat_file("trip_records/" + file)
    if file == "2023-12.parquet":
        sampled_data = sample.copy()
        continue
    # print(sampled_data.head())
    sampled_data = pd.concat([sampled_data, sample])

trip_records/2023-12.parquet
trip_records/2023-6.parquet
trip_records/2023-7.parquet
trip_records/2023-5.parquet
trip_records/2023-11.parquet
trip_records/2023-10.parquet
trip_records/2023-4.parquet
trip_records/2023-1.parquet
trip_records/2023-8.parquet
trip_records/2023-9.parquet
trip_records/2023-2.parquet
trip_records/2023-3.parquet


In [90]:
sampled_data

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,...,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee,hour,airport_fee
2225435,2,2023-12-20 00:19:54,2023-12-20 00:30:46,1.0,2.55,1.0,N,143,68,1,...,1.0,0.5,3.98,0.00,1.0,23.88,2.5,0.00,0,NaN
1916769,1,2023-12-17 00:48:57,2023-12-17 00:51:54,1.0,0.80,1.0,N,170,233,1,...,3.5,0.5,3.20,0.00,1.0,14.00,2.5,0.00,0,NaN
1920448,1,2023-12-17 00:05:01,2023-12-17 00:14:14,2.0,1.20,1.0,N,68,68,1,...,3.5,0.5,3.00,0.00,1.0,18.00,2.5,0.00,0,NaN
3306208,2,2023-12-17 00:24:42,2023-12-17 00:46:16,NaN,3.01,NaN,None,79,237,0,...,0.0,0.5,5.74,0.00,1.0,34.44,NaN,NaN,0,NaN
2800410,1,2023-12-27 00:58:36,2023-12-27 01:04:13,1.0,2.00,1.0,N,79,170,1,...,3.5,0.5,3.75,0.00,1.0,18.75,2.5,0.00,0,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1119345,2,2023-06-10 23:30:39,2023-06-10 23:45:30,1.0,1.97,1.0,N,161,249,2,...,1.0,0.5,0.00,0.00,1.0,19.90,2.5,0.00,23,NaN
1552268,2,2023-06-14 23:11:21,2023-06-14 23:13:24,1.0,0.47,1.0,N,141,236,3,...,1.0,0.5,0.00,0.00,1.0,9.40,2.5,0.00,23,NaN
1317680,2,2023-06-12 23:17:58,2023-06-12 23:47:42,2.0,9.99,1.0,N,138,113,1,...,6.0,0.5,10.00,6.55,1.0,72.60,2.5,1.75,23,NaN
783656,2,2023-06-07 23:02:53,2023-06-07 23:24:18,1.0,7.01,1.0,N,161,255,1,...,1.0,0.5,8.79,6.55,1.0,52.74,2.5,0.00,23,NaN


In [91]:
sampled_data.head()

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,...,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee,hour,airport_fee
2225435,2,2023-12-20 00:19:54,2023-12-20 00:30:46,1.0,2.55,1.0,N,143,68,1,...,1.0,0.5,3.98,0.0,1.0,23.88,2.5,0.0,0,NaN
1916769,1,2023-12-17 00:48:57,2023-12-17 00:51:54,1.0,0.80,1.0,N,170,233,1,...,3.5,0.5,3.20,0.0,1.0,14.00,2.5,0.0,0,NaN
1920448,1,2023-12-17 00:05:01,2023-12-17 00:14:14,2.0,1.20,1.0,N,68,68,1,...,3.5,0.5,3.00,0.0,1.0,18.00,2.5,0.0,0,NaN
3306208,2,2023-12-17 00:24:42,2023-12-17 00:46:16,NaN,3.01,NaN,None,79,237,0,...,0.0,0.5,5.74,0.0,1.0,34.44,NaN,NaN,0,NaN
2800410,1,2023-12-27 00:58:36,2023-12-27 01:04:13,1.0,2.00,1.0,N,79,170,1,...,3.5,0.5,3.75,0.0,1.0,18.75,2.5,0.0,0,NaN


In [92]:
sampled_data.tail()

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,...,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee,hour,airport_fee
1119345,2,2023-06-10 23:30:39,2023-06-10 23:45:30,1.0,1.97,1.0,N,161,249,2,...,1.0,0.5,0.00,0.00,1.0,19.90,2.5,0.00,23,NaN
1552268,2,2023-06-14 23:11:21,2023-06-14 23:13:24,1.0,0.47,1.0,N,141,236,3,...,1.0,0.5,0.00,0.00,1.0,9.40,2.5,0.00,23,NaN
1317680,2,2023-06-12 23:17:58,2023-06-12 23:47:42,2.0,9.99,1.0,N,138,113,1,...,6.0,0.5,10.00,6.55,1.0,72.60,2.5,1.75,23,NaN
783656,2,2023-06-07 23:02:53,2023-06-07 23:24:18,1.0,7.01,1.0,N,161,255,1,...,1.0,0.5,8.79,6.55,1.0,52.74,2.5,0.00,23,NaN
346948,1,2023-06-03 23:05:50,2023-06-03 23:22:06,1.0,2.00,1.0,N,100,158,1,...,3.5,0.5,4.65,0.00,1.0,23.15,2.5,0.00,23,NaN


In [95]:
check = concat_file("trip_records/2023-1.parquet")
check

trip_records/2023-1.parquet


,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,airport_fee,hour
784218,2,2023-01-10 00:44:09,2023-01-10 00:51:23,1.0,1.97,1.0,N,142,164,1,10.7,1.00,0.5,3.14,0.0,1.0,18.84,2.5,0.00,0
1316921,1,2023-01-15 00:32:10,2023-01-15 00:42:27,1.0,1.50,1.0,N,113,79,1,10.0,3.50,0.5,2.00,0.0,1.0,17.00,2.5,0.00,0
2732390,1,2023-01-29 00:29:35,2023-01-29 00:42:50,2.0,2.60,1.0,N,263,143,1,15.6,3.50,0.5,4.10,0.0,1.0,24.70,2.5,0.00,0
1203249,2,2023-01-14 00:14:30,2023-01-14 00:27:04,1.0,1.79,1.0,N,79,158,1,13.5,1.00,0.5,1.48,0.0,1.0,19.98,2.5,0.00,0
137926,2,2023-01-03 00:55:05,2023-01-03 01:03:28,1.0,3.34,1.0,N,132,132,2,15.6,1.00,0.5,0.00,0.0,1.0,19.35,0.0,1.25,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1904739,2,2023-01-20 23:21:31,2023-01-20 23:28:40,1.0,0.98,1.0,N,249,113,2,8.6,1.00,0.5,0.00,0.0,1.0,13.60,2.5,0.00,23
71943,2,2023-01-01 23:16:42,2023-01-01 23:45:50,1.0,11.66,1.0,N,264,37,1,48.5,1.00,0.5,3.74,0.0,1.0,57.24,2.5,0.00,23
2397817,1,2023-01-25 23:52:02,2023-01-26 00:02:31,4.0,5.10,1.0,N,132,216,2,21.9,2.25,0.5,0.00,0.0,1.0,25.65,0.0,1.25,23
1483332,2,2023-01-16 23:50:01,2023-01-16 23:54:26,1.0,1.83,1.0,N,74,263,1,9.3,1.00,0.5,2.86,0.0,1.0,17.16,2.5,0.00,23


In [96]:
sampled_data.shape

(1896428, 21)

In [97]:
sampled_data.to_parquet("total.parquet")

How many rows are there? Do you think handling such a large number of rows is computationally feasible when we have to combine the data for all twelve months into one?

To handle this, we need to sample a fraction of data from each of the files. How to go about that? Think of a way to select only some portion of the data from each month's file that accurately represents the trends.

#### Sampling the Data
> One way is to take a small percentage of entries for pickup in every hour of a date. So, for all the days in a month, we can iterate through the hours and select 5% values randomly from those. Use `tpep_pickup_datetime` for this. Separate date and hour from the datetime values and then for each date, select some fraction of trips for each of the 24 hours.

To sample data, you can use the `sample()` method. Follow this syntax:

```Python
# sampled_data is an empty DF to keep appending sampled data of each hour
# hour_data is the DF of entries for an hour 'X' on a date 'Y'

sample = hour_data.sample(frac = 0.05, random_state = 42)
# sample 0.05 of the hour_data
# random_state is just a seed for sampling, you can define it yourself

sampled_data = pd.concat([sampled_data, sample]) # adding data for this hour to the DF
```

This *sampled_data* will contain 5% values selected at random from each hour.

Note that the code given above is only the part that will be used for sampling and not the complete code required for sampling and combining the data files.

Keep in mind that you sample by date AND hour, not just hour. (Why?)

---

**1.1.1** <font color = red>[5 marks]</font> <br>
Figure out how to sample and combine the files.

**Note:** It is not mandatory to use the method specified above. While sampling, you only need to make sure that your sampled data represents the overall data of all the months accurately.

In [ ]:
# Sample the data
# It is recommmended to not load all the files at once to avoid memory overload

In [ ]:
# from google.colab import drive
# drive.mount('/content/drive')

In [ ]:
# Take a small percentage of entries from each hour of every date.
# Iterating through the monthly data:
#   read a month file -> day -> hour: append sampled data -> move to next hour -> move to next day after 24 hours -> move to next month file
# Create a single dataframe for the year combining all the monthly data

# Select the folder having data files
import os

# Select the folder having data files
os.chdir('/content/Assignments/EDA/data_NYC_Taxi/trip_records')

# Create a list of all the twelve files to read
file_list = os.listdir()

# initialise an empty dataframe
df = pd.DataFrame()


# iterate through the list of files and sample one by one:
for file_name in file_list:
    try:
        # file path for the current file
        file_path = os.path.join(os.getcwd(), file_name)

        # Reading the current file



        # We will store the sampled data for the current date in this df by appending the sampled data from each hour to this
        # After completing iteration through each date, we will append this data to the final dataframe.
        sampled_data = pd.DataFrame()

        # Loop through dates and then loop through every hour of each date

            # Iterate through each hour of the selected date

                # Sample 5% of the hourly data randomly

                # add data of this hour to the dataframe

        # Concatenate the sampled data of all the dates to a single dataframe
        df = # we initialised this empty DF earlier

    except Exception as e:
        print(f"Error reading file {file_name}: {e}")

After combining the data files into one DataFrame, convert the new DataFrame to a CSV or parquet file and store it to use directly.

Ideally, you can try keeping the total entries to around 250,000 to 300,000.

In [ ]:
# Store the df in csv/parquet
# df.to_parquet('')

## **2** Data Cleaning
<font color = red>[30 marks]</font> <br>

Now we can load the new data directly.

In [ ]:
# Load the new data file

In [ ]:
# df.head()

In [ ]:
# df.info()

#### **2.1** Fixing Columns
<font color = red>[10 marks]</font> <br>

Fix/drop any columns as you seem necessary in the below sections

**2.1.1** <font color = red>[2 marks]</font> <br>

Fix the index and drop unnecessary columns

In [ ]:
# Fix the index and drop any columns that are not needed

**2.1.2** <font color = red>[3 marks]</font> <br>
There are two airport fee columns. This is possibly an error in naming columns. Let's see whether these can be combined into a single column.

In [ ]:
# Combine the two airport fee columns

**2.1.3** <font color = red>[5 marks]</font> <br>
Fix columns with negative (monetary) values

In [ ]:
# check where values of fare amount are negative

Did you notice something different in the `RatecodeID` column for above records?

In [ ]:
# Analyse RatecodeID for the negative fare amounts

In [ ]:
# Find which columns have negative values

In [ ]:
# fix these negative values

### **2.2** Handling Missing Values
<font color = red>[10 marks]</font> <br>

**2.2.1**  <font color = red>[2 marks]</font> <br>
Find the proportion of missing values in each column




In [ ]:
# Find the proportion of missing values in each column

**2.2.2**  <font color = red>[3 marks]</font> <br>
Handling missing values in `passenger_count`

In [ ]:
# Display the rows with null values
# Impute NaN values in 'passenger_count'

Did you find zeroes in passenger_count? Handle these.

**2.2.3**  <font color = red>[2 marks]</font> <br>
Handle missing values in `RatecodeID`

In [ ]:
# Fix missing values in 'RatecodeID'

**2.2.4**  <font color = red>[3 marks]</font> <br>
Impute NaN in `congestion_surcharge`

In [ ]:
# handle null values in congestion_surcharge

Are there missing values in other columns? Did you find NaN values in some other set of columns? Handle those missing values below.

In [ ]:
# Handle any remaining missing values

### **2.3** Handling Outliers
<font color = red>[10 marks]</font> <br>

Before we start fixing outliers, let's perform outlier analysis.

In [ ]:
# Describe the data and check if there are any potential outliers present
# Check for potential out of place values in various columns

**2.3.1**  <font color = red>[10 marks]</font> <br>
Based on the above analysis, it seems that some of the outliers are present due to errors in registering the trips. Fix the outliers.

Some points you can look for:
- Entries where `trip_distance` is nearly 0 and `fare_amount` is more than 300
- Entries where `trip_distance` and `fare_amount` are 0 but the pickup and dropoff zones are different (both distance and fare should not be zero for different zones)
- Entries where `trip_distance` is more than 250  miles.
- Entries where `payment_type` is 0 (there is no payment_type 0 defined in the data dictionary)

These are just some suggestions. You can handle outliers in any way you wish, using the insights from above outlier analysis.

How will you fix each of these values? Which ones will you drop and which ones will you replace?

First, let us remove 7+ passenger counts as there are very less instances.

In [ ]:
# remove passenger_count > 6

In [ ]:
# Continue with outlier handling

In [ ]:
# Do any columns need standardising?

## **3** Exploratory Data Analysis
<font color = red>[90 marks]</font> <br>

In [ ]:
df.columns.tolist()

#### **3.1** General EDA: Finding Patterns and Trends
<font color = red>[40 marks]</font> <br>

**3.1.1** <font color = red>[3 marks]</font> <br>
Categorise the varaibles into Numerical or Categorical.
* `VendorID`:
* `tpep_pickup_datetime`:
* `tpep_dropoff_datetime`:
* `passenger_count`:
* `trip_distance`:
* `RatecodeID`:
* `PULocationID`:
* `DOLocationID`:
* `payment_type`:
* `pickup_hour`:
* `trip_duration`:


The following monetary parameters belong in the same category, is it categorical or numerical?


* `fare_amount`
* `extra`
* `mta_tax`
* `tip_amount`
* `tolls_amount`
* `improvement_surcharge`
* `total_amount`
* `congestion_surcharge`
* `airport_fee`

##### Temporal Analysis

**3.1.2** <font color = red>[5 marks]</font> <br>
Analyse the distribution of taxi pickups by hours, days of the week, and months.

In [ ]:
# Find and show the hourly trends in taxi pickups

In [ ]:
# Find and show the daily trends in taxi pickups (days of the week)

In [ ]:
# Show the monthly trends in pickups

##### Financial Analysis

Take a look at the financial parameters like `fare_amount`, `tip_amount`, `total_amount`, and also `trip_distance`. Do these contain zero/negative values?

In [ ]:
# Analyse the above parameters

Do you think it is beneficial to create a copy DataFrame leaving out the zero values from these?

**3.1.3** <font color = red>[2 marks]</font> <br>
Filter out the zero values from the above columns.

**Note:** The distance might be 0 in cases where pickup and drop is in the same zone. Do you think it is suitable to drop such cases of zero distance?

In [ ]:
# Create a df with non zero entries for the selected parameters.

**3.1.4** <font color = red>[3 marks]</font> <br>
Analyse the monthly revenue (`total_amount`) trend

In [ ]:
# Group data by month and analyse monthly revenue

**3.1.5** <font color = red>[3 marks]</font> <br>
Show the proportion of each quarter of the year in the revenue

In [ ]:
# Calculate proportion of each quarter

**3.1.6** <font color = red>[3 marks]</font> <br>
Visualise the relationship between `trip_distance` and `fare_amount`. Also find the correlation value for these two.

**Hint:** You can leave out the trips with trip_distance = 0

In [ ]:
# Show how trip fare is affected by distance

**3.1.7** <font color = red>[5 marks]</font> <br>
Find and visualise the correlation between:
1. `fare_amount` and trip duration (pickup time to dropoff time)
2. `fare_amount` and `passenger_count`
3. `tip_amount` and `trip_distance`

In [ ]:
# Show relationship between fare and trip duration

In [ ]:
# Show relationship between fare and number of passengers

In [ ]:
# Show relationship between tip and trip distance

**3.1.8** <font color = red>[3 marks]</font> <br>
Analyse the distribution of different payment types (`payment_type`)

In [ ]:
# Analyse the distribution of different payment types (payment_type).

- 1= Credit card
- 2= Cash
- 3= No charge
- 4= Dispute



##### Geographical Analysis

For this, you have to use the *taxi_zones.shp* file from the *taxi_zones* folder.

There would be multiple files inside the folder (such as *.shx, .sbx, .sbn* etc). You do not need to import/read any of the files other than the shapefile, *taxi_zones.shp*.

Do not change any folder structure - all the files need to be present inside the folder for it to work.

The folder structure should look like this:
```
Taxi Zones
|- taxi_zones.shp.xml
|- taxi_zones.prj
|- taxi_zones.sbn
|- taxi_zones.shp
|- taxi_zones.dbf
|- taxi_zones.shx
|- taxi_zones.sbx

 ```

 You only need to read the `taxi_zones.shp` file. The *shp* file will utilise the other files by itself.

We will use the *GeoPandas* library for geopgraphical analysis
```
import geopandas as gpd
```

More about geopandas and shapefiles: [About](https://geopandas.org/en/stable/about.html)


Reading the shapefile is very similar to *Pandas*. Use `gpd.read_file()` function to load the data (*taxi_zones.shp*) as a GeoDataFrame. Documentation: [Reading and Writing Files](https://geopandas.org/en/stable/docs/user_guide/io.html)

In [ ]:
# !pip install geopandas

**3.1.9** <font color = red>[2 marks]</font> <br>
Load the shapefile and display it.

In [ ]:
# import geopandas as gpd


# Read the shapefile using geopandas
zones = # read the .shp file using gpd
zones.head()

Now, if you look at the DataFrame created, you will see columns like: `OBJECTID`,`Shape_Leng`, `Shape_Area`, `zone`, `LocationID`, `borough`, `geometry`.
<br><br>

Now, the `locationID` here is also what we are using to mark pickup and drop zones in the trip records.

The geometric parameters like shape length, shape area and geometry are used to plot the zones on a map.

This can be easily done using the `plot()` method.

In [ ]:
# print(zones.info())
# zones.plot()

Now, you have to merge the trip records and zones data using the location IDs.



**3.1.10** <font color = red>[3 marks]</font> <br>
Merge the zones data into trip data using the `locationID` and `PULocationID` columns.

In [ ]:
# Merge zones and trip records using locationID and PULocationID

**3.1.11** <font color = red>[3 marks]</font> <br>
Group data by location IDs to find the total number of trips per location ID

In [ ]:
# Group data by location and calculate the number of trips

**3.1.12** <font color = red>[2 marks]</font> <br>
Now, use the grouped data to add number of trips to the GeoDataFrame.

We will use this to plot a map of zones showing total trips per zone.

In [ ]:
# Merge trip counts back to the zones GeoDataFrame

The next step is creating a color map (choropleth map) showing zones by the number of trips taken.

Again, you can use the `zones.plot()` method for this. [Plot Method GPD](https://geopandas.org/en/stable/docs/reference/api/geopandas.GeoDataFrame.plot.html#geopandas.GeoDataFrame.plot)

But first, you need to define the figure and axis for the plot.

`fig, ax = plt.subplots(1, 1, figsize = (12, 10))`

This function creates a figure (fig) and a single subplot (ax)

---

After setting up the figure and axis, we can proceed to plot the GeoDataFrame on this axis. This is done in the next step where we use the plot method of the GeoDataFrame.

You can define the following parameters in the `zones.plot()` method:
```
column = '',
ax = ax,
legend = True,
legend_kwds = {'label': "label", 'orientation': "<horizontal/vertical>"}
```

To display the plot, use `plt.show()`.

**3.1.13** <font color = red>[3 marks]</font> <br>
Plot a color-coded map showing zone-wise trips

In [ ]:
# Define figure and axis


# Plot the map and display it

In [ ]:
# can you try displaying the zones DF sorted by the number of trips?

Here we have completed the temporal, financial and geographical analysis on the trip records.

**Compile your findings from general analysis below:**

You can consider the following points:

* Busiest hours, days and months
* Trends in revenue collected
* Trends in quarterly revenue
* How fare depends on trip distance, trip duration and passenger counts
* How tip amount depends on trip distance
* Busiest zones


#### **3.2** Detailed EDA: Insights and Strategies
<font color = red>[50 marks]</font> <br>

Having performed basic analyses for finding trends and patterns, we will now move on to some detailed analysis focussed on operational efficiency, pricing strategies, and customer experience.

##### Operational Efficiency

Analyze variations by time of day and location to identify bottlenecks or inefficiencies in routes

**3.2.1** <font color = red>[3 marks]</font> <br>
Identify slow routes by calculating the average time taken by cabs to get from one zone to another at different hours of the day.

Speed on a route *X* for hour *Y* = (*distance of the route X / average trip duration for hour Y*)

In [ ]:
# Find routes which have the slowest speeds at different times of the day

How does identifying high-traffic, high-demand routes help us?

**3.2.2** <font color = red>[3 marks]</font> <br>
Calculate the number of trips at each hour of the day and visualise them. Find the busiest hour and show the number of trips for that hour.

In [ ]:
# Visualise the number of trips per hour and find the busiest hour

Remember, we took a fraction of trips. To find the actual number, you have to scale the number up by the sampling ratio.

**3.2.3** <font color = red>[2 mark]</font> <br>
Find the actual number of trips in the five busiest hours

In [ ]:
# Scale up the number of trips

# Fill in the value of your sampling fraction and use that to scale up the numbers
sample_fraction =



**3.2.4** <font color = red>[3 marks]</font> <br>
Compare hourly traffic pattern on weekdays. Also compare for weekend.

In [ ]:
# Compare traffic trends for the week days and weekends

What can you infer from the above patterns? How will finding busy and quiet hours for each day help us?

**3.2.5** <font color = red>[3 marks]</font> <br>
Identify top 10 zones with high hourly pickups. Do the same for hourly dropoffs. Show pickup and dropoff trends in these zones.

In [ ]:
# Find top 10 pickup and dropoff zones

**3.2.6** <font color = red>[3 marks]</font> <br>
Find the ratio of pickups and dropoffs in each zone. Display the 10 highest (pickup/drop) and 10 lowest (pickup/drop) ratios.

In [ ]:
# Find the top 10 and bottom 10 pickup/dropoff ratios

**3.2.7** <font color = red>[3 marks]</font> <br>
Identify zones with high pickup and dropoff traffic during night hours (11PM to 5AM)

In [ ]:
# During night hours (11pm to 5am) find the top 10 pickup and dropoff zones
# Note that the top zones should be of night hours and not the overall top zones

Now, let us find the revenue share for the night time hours and the day time hours. After this, we will move to deciding a pricing strategy.

**3.2.8** <font color = red>[2 marks]</font> <br>
Find the revenue share for nighttime and daytime hours.

In [ ]:
# Filter for night hours (11 PM to 5 AM)

##### Pricing Strategy

**3.2.9** <font color = red>[2 marks]</font> <br>
For the different passenger counts, find the average fare per mile per passenger.

For instance, suppose the average fare per mile for trips with 3 passengers is 3 USD/mile, then the fare per mile per passenger will be 1 USD/mile.

In [ ]:
# Analyse the fare per mile per passenger for different passenger counts

**3.2.10** <font color = red>[3 marks]</font> <br>
Find the average fare per mile by hours of the day and by days of the week

In [ ]:
# Compare the average fare per mile for different days and for different times of the day

**3.2.11** <font color = red>[3 marks]</font> <br>
Analyse the average fare per mile for the different vendors for different hours of the day

In [ ]:
# Compare fare per mile for different vendors

**3.2.12** <font color = red>[5 marks]</font> <br>
Compare the fare rates of the different vendors in a tiered fashion. Analyse the average fare per mile for distances upto 2 miles. Analyse the fare per mile for distances from 2 to 5 miles. And then for distances more than 5 miles.


In [ ]:
# Defining distance tiers

##### Customer Experience and Other Factors

**3.2.13** <font color = red>[5 marks]</font> <br>
Analyse average tip percentages based on trip distances, passenger counts and time of pickup. What factors lead to low tip percentages?

In [ ]:
#  Analyze tip percentages based on distances, passenger counts and pickup times

Additional analysis [optional]: Let's try comparing cases of low tips with cases of high tips to find out if we find a clear aspect that drives up the tipping behaviours

In [ ]:
# Compare trips with tip percentage < 10% to trips with tip percentage > 25%

**3.2.14** <font color = red>[3 marks]</font> <br>
Analyse the variation of passenger count across hours and days of the week.

In [ ]:
# See how passenger count varies across hours and days

**3.2.15** <font color = red>[2 marks]</font> <br>
Analyse the variation of passenger counts across zones

In [ ]:
# How does passenger count vary across zones

In [ ]:
# For a more detailed analysis, we can use the zones_with_trips GeoDataFrame
# Create a new column for the average passenger count in each zone.

Find out how often surcharges/extra charges are applied to understand their prevalance

**3.2.16** <font color = red>[5 marks]</font> <br>
Analyse the pickup/dropoff zones or times when extra charges are applied more frequently

In [ ]:
# How often is each surcharge applied?

## **4** Conclusion
<font color = red>[15 marks]</font> <br>

### **4.1** Final Insights and Recommendations
<font color = red>[15 marks]</font> <br>

Conclude your analyses here. Include all the outcomes you found based on the analysis.

Based on the insights, frame a concluding story explaining suitable parameters such as location, time of the day, day of the week etc. to be kept in mind while devising a strategy to meet customer demand and optimise supply.

**4.1.1** <font color = red>[5 marks]</font> <br>
Recommendations to optimize routing and dispatching based on demand patterns and operational inefficiencies

**4.1.2** <font color = red>[5 marks]</font> <br>

Suggestions on strategically positioning cabs across different zones to make best use of insights uncovered by analysing trip trends across time, days and months.

**4.1.3** <font color = red>[5 marks]</font> <br>
Propose data-driven adjustments to the pricing strategy to maximize revenue while maintaining competitive rates with other vendors.